### ISSUE TITLE
[IDE0004] False positive for null literal in interpolated string

### ISSUE THREAD

**Issue by MaceWindu** [dotnet/roslyn#75145]
**Version Used**: 4.11.0-3.24428.4+1ea9c390a5bb6815fdff2137ee155e23e78d6ff3

**Steps to Reproduce**:

```xml
<PropertyGroup>
  <RunAnalyzersDuringBuild>true</RunAnalyzersDuringBuild>
  <RunAnalyzersDuringLiveAnalysis>true</RunAnalyzersDuringLiveAnalysis>
  <AnalysisLevel>preview-All</AnalysisLevel>
  <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
</PropertyGroup>
```

```cs
using System.Runtime.CompilerServices;

public class IssueClass
{
	public object? Test()
	{
		// CS0411: The type arguments for method '...' cannot be inferred from the usage. Try specifying the type arguments explicitly
		//return InterpolateMe($"{null}");
                // IDE0004: Cast is redundant.
		return InterpolateMe($"{(object?)null}");
	}

	public static object? InterpolateMe([InterpolatedStringHandlerArgument] ref TestInterpolatedStringHandler handler)
	{
		return null;
	}

	[InterpolatedStringHandler]
	public ref struct TestInterpolatedStringHandler(int literalLength, int formattedCount)
	{
		public readonly void AppendLiteral(string s)
		{
		}

		public void AppendFormatted<T>(T value)
		{
		}
	}
}

```

**Diagnostic Id**:

IDE0004	Cast is redundant. (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0004)

**Expected Behavior**:

no IDE0004 reported

**Actual Behavior**:

IDE0004 reported

**Comment by victor-pogor** [comment]
I took as an example the `DefaultInterpolatedStringHandler`, there is no error if removing the redundant cast.

```diff
using System;
public class C {
    public static void Main() {
-       string s = $"{(object?)null}";
+       string s = $"{null}";
        Console.WriteLine(s);
    }
}
```

This is the code diff:

```diff
public class C
{
    public static void Main()
    {
        DefaultInterpolatedStringHandler defaultInterpolatedStringHandler = new DefaultInterpolatedStringHandler(0, 1);
-       defaultInterpolatedStringHandler.AppendFormatted<object>(null);
+       defaultInterpolatedStringHandler.AppendFormatted(null);
        string value = defaultInterpolatedStringHandler.ToStringAndClear();
        Console.WriteLine(value);
    }
}
```

There is no error because `DefaultInterpolatedStringHandler` has an overload for [AppendFormatted](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.defaultinterpolatedstringhandler.appendformatted?view=net-8.0#system-runtime-compilerservices-defaultinterpolatedstringhandler-appendformatted(system-object-system-int32-system-string)) method:

```c#
public void AppendFormatted (object? value, int alignment = 0, string? format = default);
```

If adding the same overload to the custom handler, there is no error:

```diff
using System.Runtime.CompilerServices;

public class IssueClass
{
	public object? Test()
	{
+               // Works in this case
		return InterpolateMe($"{null}");
	}

	public static object? InterpolateMe([InterpolatedStringHandlerArgument] ref TestInterpolatedStringHandler handler)
	{
		return null;
	}

	[InterpolatedStringHandler]
	public ref struct TestInterpolatedStringHandler(int literalLength, int formattedCount)
	{
		public readonly void AppendLiteral(string s)
		{
		}

		public void AppendFormatted<T>(T value)
		{
		}

+               // Adding a similar overload as DefaultInterpolatedStringHandler
+               public void AppendFormatted(object? value)
+               {
+               }
	}
}
```

I guess the magic happens in [CastSimplifier](https://github.com/dotnet/roslyn/blob/54fd790788b03c256f736c917fc368daf0a643db/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs) class.

https://github.com/dotnet/roslyn/blob/54fd790788b03c256f736c917fc368daf0a643db/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs#L280-L291

I tried to look in the overload resolution method but not sure is the right place

![image](https://github.com/user-attachments/assets/1e9422f9-30c4-4c9a-81f1-917f9d6030c2)

Is there a way to check whether this is a custom interpolation handler and not showing this diagnostic? Not sure when the interpolation handler gets converted to full code.

Is there a better way?
